ludwig/utils/fs_utils.py

```diff
@@ -0,0 +1,123 @@
+#! /usr/bin/env python
+# coding=utf-8
+# Copyright (c) 2021 Linux Foundation.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+# ==============================================================================
+
+import contextlib
+import os
+import pathlib
+import tempfile
+
+import fsspec
+import h5py
+from fsspec.core import split_protocol
+
+
+def get_fs_and_path(url):
+    protocol, path = split_protocol(url)
+    fs = fsspec.filesystem(protocol)
+    return fs, path
+
+
+def find_non_existing_dir_by_adding_suffix(directory_name):
+    fs, _ = get_fs_and_path(directory_name)
+    suffix = 0
+    curr_directory_name = directory_name
+    while fs.exists(curr_directory_name):
+        curr_directory_name = directory_name + '_' + str(suffix)
+        suffix += 1
+    return curr_directory_name
+
+
+def path_exists(url):
+    fs, path = get_fs_and_path(url)
+    return fs.exists(path)
+
+
+def makedirs(url, exist_ok=False):
+    fs, path = get_fs_and_path(url)
+    return fs.makedirs(path, exist_ok=exist_ok)
+
+
+def delete(url, recursive=False):
+    fs, path = get_fs_and_path(url)
+    return fs.delete(path, recursive=recursive)
+
+
+def checksum(url):
+    fs, path = get_fs_and_path(url)
+    return fs.checksum(path)
+
+
+def to_url(path):
+    protocol, _ = split_protocol(path)
+    if protocol is not None:
+        return path
+    return pathlib.Path(os.path.abspath(path)).as_uri()
+
+
+@contextlib.contextmanager
+def upload_output_directory(url):
+    if url is None:
+        yield None
+        return
+
+    protocol, _ = split_protocol(url)
+    if protocol is not None:
+        # To avoid extra network load, write all output files locally at runtime,
+        # then upload to the remote fs at the end.
+        with tempfile.TemporaryDirectory() as tmpdir:
+            fs, remote_path = get_fs_and_path(url)
+            if path_exists(url):
+                fs.get(url, tmpdir + '/', recursive=True)
+
+            # Write to temp directory locally
+            yield tmpdir
+
+            # Upload to remote when finished
+            fs.put(tmpdir, remote_path, recursive=True)
+    else:
+        # Just use the output directory directly if using a local filesystem
+        yield url
+
+
+@contextlib.contextmanager
+def open_file(url, *args, **kwargs):
+    of = fsspec.open(url, *args, **kwargs)
+    with of as f:
+        yield f
+
+
+@contextlib.contextmanager
+def download_h5(url):
```

Comment:
why download_h5 and not open_h5?

Reply:
also does this work with remote urls? I guess if it's remote fsspec.open_local downloads it first?

Reply:
`download_h5` will first download the remote file to local, then open it for read-only.

`upload_h5` opens the file for write then uploads to the remote url.

The `open_local` just means to download the file locally before opening, yes, as opposed to streaming. This is primarily for the purpose of caching (if you intend to re-read repeatedly).